lib/teleterm/gateway/gateway.go

```diff
@@ -70,55 +70,25 @@ func New(cfg Config) (*Gateway, error) {
 
 	cfg.LocalPort = port
 
-	tlsCert, err := keys.LoadX509KeyPair(cfg.CertPath, cfg.KeyPath)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	if err := checkCertSubject(tlsCert, cfg.RouteToDatabase()); err != nil {
-		return nil, trace.Wrap(err,
-			"database certificate check failed, try restarting the database connection")
-	}
-
-	localProxyConfig := alpn.LocalProxyConfig{
-		InsecureSkipVerify:      cfg.Insecure,
-		RemoteProxyAddr:         cfg.WebProxyAddr,
-		Listener:                listener,
-		ParentContext:           closeContext,
-		Certs:                   []tls.Certificate{tlsCert},
-		Clock:                   cfg.Clock,
-		ALPNConnUpgradeRequired: cfg.TLSRoutingConnUpgradeRequired,
-	}
-
-	localProxyMiddleware := &localProxyMiddleware{
-		log:     cfg.Log,
-		dbRoute: cfg.RouteToDatabase(),
-	}
-
-	if cfg.OnExpiredCert != nil {
-		localProxyConfig.Middleware = localProxyMiddleware
-	}
-
-	localProxy, err := alpn.NewLocalProxy(localProxyConfig,
-		alpn.WithDatabaseProtocol(cfg.Protocol),
-		alpn.WithClusterCAsIfConnUpgrade(closeContext, cfg.RootClusterCACertPoolFunc),
-	)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
 	gateway := &Gateway{
 		cfg:          &cfg,
 		closeContext: closeContext,
 		closeCancel:  closeCancel,
-		localProxy:   localProxy,
 	}
 
-	if cfg.OnExpiredCert != nil {
-		localProxyMiddleware.onExpiredCert = func(ctx context.Context) error {
-			err := cfg.OnExpiredCert(ctx, gateway)
-			return trace.Wrap(err)
+	switch targetURI := uri.New(cfg.TargetURI); {
+	case targetURI.IsDB():
+		if err := gateway.makeLocalProxyForDB(listener); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+	case targetURI.IsKube():
+		if err := gateway.makeLocalProxiesForKube(listener); err != nil {
+			return nil, trace.Wrap(err)
```

Comment:
Cluster has to issue different certs and pass different parameters for kube vs db. So it's hard to avoid this check on separate levels. 

An alternative that makes it slightly easier to read is to define a gateway type. but it doesn't eliminate the switch tho so i didn't do it.